src/boot/efi/util.c

```diff
@@ -501,3 +501,15 @@ EFI_STATUS log_oom(void) {
         log_error_stall(L"Out of memory.");
         return EFI_OUT_OF_RESOURCES;
 }
+
+VOID *memmem_safe(const VOID *haystack, UINTN haystack_len, const VOID *needle, UINTN needle_len) {
+        assert(haystack);
+        assert(needle);
+        assert(needle_len > 0);
```

Comment:
I'd model this after memmem_safe() in userspace, i.e. allow zero length haystack and needle.

i.e. add two early checks here:
```c

if (needle_len == 0)
        return haystack;
if (haystack_len == needle_len)
        return NULL;
```